### PR TITLE
Bubble sort

### PR DESCRIPTION
**Pull request by oleksandr-maksymikhin**
This pull request includes implementation of Bubble sort algorithm:

1.	Implementation of bubble sort algorithm is in solutions/bubble_sort.py
2.	Unit tests are in solutions/tests/ test_bubble_sort.py
3.	Passed automation checks (pydoc, doctest, unittest)

Please review the solution and let me know if any corrections are needed.

### CODE REVIEW

File: solutions/bubble_sort.py
Comment (emrebiyik) [inline solutions/bubble_sort.py:1]: Suggestions for Improvement:
Type hints in Python provide additional information about the types of variables, function arguments, and return values. They help developers understand the expected inputs and outputs of a function, making the code more readable and less error-prone.

In the context of bubble_sort.py, here are two key suggestions to enhance clarity and robustness:
1. Adding Type Hints:
• Currently, the function accepts list[any], which means it can take a list containing elements of any type. However, not all types are comparable. For example, comparing an int with a str will raise a TypeError.
• Adding specific type hints, such as List[int], or using a generic comparable type ensures that the function only accepts lists of elements that can be compared.
2. Ensuring Comparability:
• The function should guarantee that elements in the list are comparable, i.e., they support operators like <, >, and ==. Using Python’s TypeVar and Comparable classes can enforce this constraint.

Comment (emrebiyik) [reply solutions/bubble_sort.py:1]: Suggestion: Add a performance test for large datasets
To ensure the algorithm is scalable, I recommend adding a performance test for larger datasets (e.g., 10,000+ elements).

File: solutions/tests/test_bubble_sort.py
Comment (emrebiyik) [inline solutions/tests/test_bubble_sort.py:1]: Suggestions for Improvement:
• Add a performance test with a large dataset (e.g., 10,000+ elements) to ensure efficiency.
• Include a test for lists with identical elements to validate stability

Comment (emrebiyik) [reply solutions/tests/test_bubble_sort.py:1]: The implementation is clean, easy to follow, and well-structured. Defensive programming with assertions ensures input validation and error handling, while creating a copy of the input list avoids side effects. The test suite is thorough, covering various scenarios, including edge cases like empty and single-element lists. Great attention to detail and solid work overall!

File: solutions/bubble_sort.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+Module contents:
+    - bubble_sort: sort the collection in ascending order.
+
+Created on 2024-01-02
+Author: Oleksandr Maksymikhin
+"""
+
+
+# def bubble_sort(input_collection: list[int]) -> list[int]:
+def bubble_sort(input_collection: list[any]) -> list[any]:
```
Comment (emrebiyik) [inline solutions/bubble_sort.py]: Type hints in Python provide additional information about the types of variables, function arguments, and return values. They help developers understand the expected inputs and outputs of a function, making the code more readable and less error-prone.

In the context of bubble_sort.py, here are two key suggestions to enhance clarity and robustness:

Adding Type Hints:
• Currently, the function accepts list[any], which means it can take a list containing elements of any type. However, not all types are comparable. For example, comparing an int with a str will raise a TypeError.
• Adding specific type hints, such as List[int], or using a generic comparable type ensures that the function only accepts lists of elements that can be compared.
Ensuring Comparability:
• The function should guarantee that elements in the list are comparable, i.e., they support operators like <, >, and ==. Using Python’s TypeVar and Comparable classes can enforce this constraint.

File: solutions/tests/test_bubble_sort.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    def test_10_non_collection_input(self):
+        """It should raise an assertion error if the input is not a collection"""
+        with self.assertRaises(AssertionError):
+            bubble_sort("apple")
+
+    def test_11_non_homogeneous_collection_input(self):
+        """It should raise an assertion error if the collection is non-homogeneous"""
+        with self.assertRaises(AssertionError):
+            bubble_sort([3, "two", 1])
```
Comment (emrebiyik) [inline solutions/tests/test_bubble_sort.py:89]: Suggestions for Improvement:
• Add a performance test with a large dataset (e.g., 10,000+ elements) to ensure efficiency.
• Include a test for lists with identical elements to validate stability

**Review by emrebiyik — CHANGES_REQUESTED** [review]

File: solutions/bubble_sort.py
Comment (oleksandr-maksymikhin) [reply solutions/bubble_sort.py:1]: 1. Thanks for your suggestion with TypeVar. I know how generic data types works in statically typed programming languages (C++, C#) but I am absolutely new in Python and in interpreted programming languages at all that is why I was not able initially to resolve the issue with typing of input data. TypeVar was added with limiting of input types to int, float, str, bool.
2. Since I have limited input types to build-in elementary Python data types there is no need to add comparators to compare user’s data types.

File: solutions/tests/test_bubble_sort.py
Comment (oleksandr-maksymikhin) [reply solutions/tests/test_bubble_sort.py:1]: 1. Our class assignment and study syllabus foresee only unit testing. That is the reason why I have provided only unit tests without stability tests for huge sets of input data. I propose postpone stability tests for future stages of solution development.

2. In my test module there are tests for collection with identical and non-identical elements:

> 2.1) test for identical elements:
> test_3_two_int_elements_list
> test_4_three_int_elements_list
> test_5_four_int_elements_list_big_number
> test_6_char_elements_list
> test_7_string_elements_list
> test_8_bool_elements_list
> 
> 2.2) assertion test for non-identical elements (non-homogeneous input collection):
>  test_11_non_homogeneous_collection_input

File: solutions/tests/test_bubble_sort.py (hunk 2 above)
Comment (oleksandr-maksymikhin) [reply solutions/tests/test_bubble_sort.py:89]: 1. Our class assignment and study syllabus foresee only unit testing. That is the reason why I have provided only unit tests without stability tests for huge sets of input data. I propose postpone stability tests for future stages of solution development.

2. In my test module there are tests for collection with identical and non-identical elements:

> 2.1) test for identical elements:
> test_3_two_int_elements_list
> test_4_three_int_elements_list
> test_5_four_int_elements_list_big_number
> test_6_char_elements_list
> test_7_string_elements_list
> test_8_bool_elements_list
> 
> 2.2) assertion test for non-identical elements (non-homogeneous input collection):
> test_11_non_homogeneous_collection_input

File: solutions/bubble_sort.py
Comment (emrebiyik) [reply solutions/bubble_sort.py:1]: > 1. Thanks for your suggestion with TypeVar. I know how generic data types works in statically typed programming languages (C++, C#) but I am absolutely new in Python and in interpreted programming languages at all that is why I was not able initially to resolve the issue with typing of input data. TypeVar was added with limiting of input types to int, float, str, bool.
> 2. Since I have limited input types to build-in elementary Python data types there is no need to add comparators to compare user’s data types.

Thank you for considering my suggestion and implementing TypeVar to address the issue. Limiting input types to int, float, str, and bool is a great approach and effectively reduces the risk of TypeError.

For more complex scenarios (e.g., working with user-defined data types), I believe leveraging the Comparable class might still be a useful option in the future. However, for the current implementation, this seems unnecessary and your approach works perfectly.

Your work is truly impressive, especially considering you're new to Python.

File: solutions/tests/test_bubble_sort.py
Comment (emrebiyik) [reply solutions/tests/test_bubble_sort.py:1]: Thank you for the detailed explanation and great work! The current test coverage is impressive, and postponing stability tests for future development stages makes sense.

**Review by emrebiyik — APPROVED** [review]
Looks great! Well done, and thank you for your effort. Approved!
